### PR TITLE
Add Traffic Ops Golang ats meta config route

### PR DESCRIPTION
**Pull request by rob05c**
## What does this PR (Pull Request) do?

This adds the ATS config meta route to TO Go.

Specifically, it separates it out of https://github.com/apache/trafficcontrol/pull/2991

I've diffed this against the Perl for every production server we have, and they all diff identical.

- [x] This PR is not related to any Issue

Includes API Tests. No documentation, no changelog, not an interface change.

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

`go test`, curl `/api/1.2/servers/{id}/configfiles/ats`.

## If this is a bug fix, what versions of Traffic Ops are affected?

Not a bug fix.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3568/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3578/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3579/
Test PASSed.

**Comment by rob05c** [comment]
@limited Done

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3836/
Test FAILed.

File: traffic_ops/traffic_ops_golang/ats/meta.go [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+	qry := `
+SELECT
+  ds.xml_id
+FROM
+  deliveryservice ds
+  JOIN deliveryservice_server dss ON ds.id = dss.deliveryservice
+  JOIN server s ON s.id = dss.server
+WHERE
+  s.host_name = $1
+  AND s.tcp_port = $2
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/ats/meta.go]: We do. 

The current DB constraint is `(host, profile)`, but IMO it should really be `(host, port)`. Even with `(host,profile)`, having multiple hosts makes all the host-based API paths potentially wrong/random, we really need to address that at some point.

But for now, yeah, we need this, it'll break things for us if it doesn't.

File: traffic_ops/traffic_ops_golang/ats/meta.go [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+		return
+	}
+
+	tmParams, err := GetTMParams(inf.Tx.Tx)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("GetConfigMetaData getting tm.url parameter: "+err.Error()))
+		return
+	}
+	if tmParams.URL == "" {
+		log.Warnln("ats.GetConfigMetadata: global tm.url parameter missing or empty!")
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/ats/meta.go]: Changed. 

>What happens to ORT if this field is missing?

If nothing else, I'm pretty sure Traffic Router will break, or at least reject the CRConfig.

File: traffic_ops/traffic_ops_golang/ats/meta.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+  s.cdn_id,
+  s.host_name,
+  c.domain_name,
+  s.ip_address,
+  s.profile AS profile_id,
+  p.name AS profile_name,
+  s.tcp_port,
+  t.name as type,
+  s.cachegroup,
+  COALESCE(cg.parent_cachegroup_id, -1),
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/ats/meta.go]: Done.

It isn't technically necessary, the Go DB query doesn't care what the names are. But I agree it makes it more readable.

File: traffic_ops/traffic_ops_golang/ats/meta.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+	if locationParams["remap.config"].Location != "" {
+		configLocation := locationParams["remap.config"].Location
+		uriSignedDSes, err := GetServerURISignedDSes(inf.Tx.Tx, server.HostName, server.Port)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("GetConfigMetaData getting server uri-signed dses: "+err.Error()))
+			return
+		}
+		for _, ds := range uriSignedDSes {
+			cfgName := "uri_signing_" + string(ds) + ".config"
+			// If there's already a parameter for it, don't clobber it. The user may wish to override the location.
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/ats/meta.go]: It still should have a Parameter (and it should be auto-created). This was transliterated from Perl: https://github.com/apache/trafficcontrol/blob/56ef95e1783360b4afb600b06cf138255fb34c89/traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm#L121

I'm honestly not sure why anyone would ever want to override the file path. For URI Signing, maybe a security thing, to put it in a "locked down" location? I dunno.

File: traffic_ops/traffic_ops_golang/ats/meta.go [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+			FileNameOnDisk: cfgParams.FileNameOnDisk,
+			Location:       cfgParams.Location,
+		}
+
+		scope, err := getServerScope(inf.Tx.Tx, cfgFile, server.Type)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("GetConfigMetaData getting scope: "+err.Error()))
+			return
+		}
+		if cfgParams.URL != "" {
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/ats/meta.go]: Yep, done.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3837/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3838/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3839/
Test FAILed.

**Pull request merged**